scripts: enforce module name uniqueness

Damian-Nordic opened the pull request:
Although CMake scripts already support duplicated module names (by using ZEPHYR_${MODULE_NAME}_MODULE_DIR CMake variables and applying `list(REMOVE_DUPLICATES ZEPHYR_MODULE_NAMES)` in a different place), if two modules with the same name provide different Kconfig files, then both files will be loaded leading to potential conflicts.

Modify zephyr_module.py to enforce that all modules are uniquely named so that it is possible to override some of the built-in modules using ZEPHYR_EXTRA_MODULES variable.

Signed-off-by: Damian Krolik <damian.krolik@nordicsemi.no>

tejlmand commented on scripts/zephyr_module.py:
```diff
@@ -395,33 +398,43 @@ def parse_modules(zephyr_base, modules=None, extra_modules=None):
-            else:
-                dep_modules.append(Module(project, meta, deps))
-        elif project in extra_modules:
+        module = process_module(project)
+        if module and module.meta['name'] not in all_module_names:
+            all_modules.append(module)
+            all_module_names.add(module.meta['name'])
```
you don't need to over complicate this.

Just create an empty dictionary and use `meta` name as key.
Then you can update it based on the `meta` name, and last update will win.

Like this:
```diff
--- a/scripts/zephyr_module.py
+++ b/scripts/zephyr_module.py
@@ -396,6 +396,7 @@ def parse_modules(zephyr_base, modules=None, extra_modules=None):
         extra_modules = []
 
     Module = namedtuple('Module', ['project', 'meta', 'depends'])
+    modules_dict = {}
     # dep_modules is a list of all modules that has an unresolved dependency
     dep_modules = []
     # start_modules is a list modules with no depends left (no incoming edge)
@@ -410,16 +411,19 @@ def parse_modules(zephyr_base, modules=None, extra_modules=None):
 
         meta = process_module(project)
         if meta:
-            section = meta.get('build', dict())
-            deps = section.get('depends', [])
-            if not deps:
-                start_modules.append(Module(project, meta, []))
-            else:
-                dep_modules.append(Module(project, meta, deps))
+            modules_dict.update({meta.get('name'): {'project': project, 'meta': meta}})
         elif project in extra_modules:
             sys.exit(f'{project}, given in ZEPHYR_EXTRA_MODULES, '
                      'is not a valid zephyr module')
 
+    for name, value in modules_dict.items():
+        section = value['meta'].get('build', dict())
+        deps = section.get('depends', [])
+        if not deps:
+            start_modules.append(Module(value['project'], value['meta'], []))
+        else:
+            dep_modules.append(Module(value['project'], value['meta'], deps))
+
     # This will do a topological sort to ensure the modules are ordered
     # according to dependency settings.
     while start_modules:
```

Damian-Nordic replied:
I wanted to preserve the order or items in the module list. I was concerned that if I change it, some projects will start failing in case some modules don't have dependencies properly configured. Should I not worry? :)

tejlmand replied:
> I was concerned that if I change it, some projects will start failing in case some modules don't have dependencies properly configured. Should I not worry? :)

Correct, you should not worry.
We don't guarantee a specific order although in practice we use the order given by the argument / or provided by `west` when `west` is used.

With that said, the beauty of `dict.update()` is that it doesn't change the order, it updates at the index given.

For example with vanilla Zephyr you have this order:
```
...
10 "infineon":"/projects/github/ncs/modules/hal/infineon":"/projects/github/ncs/modules/hal/infineon"
11 "microchip":"/projects/github/ncs/modules/hal/microchip":"/projects/github/ncs/modules/hal/microchip"
12 "hal_nordic":"/projects/github/ncs/modules/hal/nordic":"${ZEPHYR_HAL_NORDIC_CMAKE_DIR}"
13 "nuvoton":"/projects/github/ncs/modules/hal/nuvoton":"/projects/github/ncs/modules/hal/nuvoton"
14 "hal_nxp":"/projects/github/ncs/modules/hal/nxp":"${ZEPHYR_HAL_NXP_CMAKE_DIR}"
...
```
and if I want to use my custom hal Nordic as: `-DZEPHYR_EXTRA_MODULES=(pwd)/../modules/hal/custom_hal_nordic/
` I'll get this:
```
...
 10 "infineon":"/projects/github/ncs/modules/hal/infineon":"/projects/github/ncs/modules/hal/infineon"
 11 "microchip":"/projects/github/ncs/modules/hal/microchip":"/projects/github/ncs/modules/hal/microchip"
 12 "hal_nordic":"/projects/github/ncs/zephyr/../modules/hal/custom_hal_nordic":"/projects/github/ncs/modules/hal/custom_hal_nordic"
 13 "nuvoton":"/projects/github/ncs/modules/hal/nuvoton":"/projects/github/ncs/modules/hal/nuvoton"
 14 "hal_nxp":"/projects/github/ncs/modules/hal/nxp":"${ZEPHYR_HAL_NXP_CMAKE_DIR}"
...
```
so as you see, order is preserved when injecting my custom version :wink:

Damian-Nordic replied:
ok, thanks. Changed it.

tejlmand reviewed (changes requested):
thanks for noticing this.

Clearly Kconfigs are not supposed to be included in this case.

But I have a small improvement proposal for the Python code.

No reason to create a named tuple, an extra list and a set and then reverse the list twice.

Just create a dict with keys == module name.
Then any update to the dict where key (name) is identical will ensure the last call wins, and as extra modules comes last then things will work as desired.

Diff with proposal (diffed against `zephyr/main`).
```diff
--- a/scripts/zephyr_module.py
+++ b/scripts/zephyr_module.py
@@ -396,6 +396,7 @@ def parse_modules(zephyr_base, modules=None, extra_modules=None):
         extra_modules = []
 
     Module = namedtuple('Module', ['project', 'meta', 'depends'])
+    modules_dict = {}
     # dep_modules is a list of all modules that has an unresolved dependency
     dep_modules = []
     # start_modules is a list modules with no depends left (no incoming edge)
@@ -410,16 +411,19 @@ def parse_modules(zephyr_base, modules=None, extra_modules=None):
 
         meta = process_module(project)
         if meta:
-            section = meta.get('build', dict())
-            deps = section.get('depends', [])
-            if not deps:
-                start_modules.append(Module(project, meta, []))
-            else:
-                dep_modules.append(Module(project, meta, deps))
+            modules_dict.update({meta.get('name'): {'project': project, 'meta': meta}})
         elif project in extra_modules:
             sys.exit(f'{project}, given in ZEPHYR_EXTRA_MODULES, '
                      'is not a valid zephyr module')
 
+    for name, value in modules_dict.items():
+        section = value['meta'].get('build', dict())
+        deps = section.get('depends', [])
+        if not deps:
+            start_modules.append(Module(value['project'], value['meta'], []))
+        else:
+            dep_modules.append(Module(value['project'], value['meta'], deps))
+
     # This will do a topological sort to ensure the modules are ordered
     # according to dependency settings.
     while start_modules:
```

Damian-Nordic committed:
scripts: enforce module name uniqueness

Although CMake scripts already support duplicated module
names, if two modules with the same name provide different
Kconfig files, then both files will be loaded leading to
potential conflicts.

Modify zephyr_module.py to enforce that all modules are
uniquely named so that it is possbile to override some of
the built-in modules using ZEPHYR_EXTRA_MODULES variable.

Signed-off-by: Damian Krolik <damian.krolik@nordicsemi.no>

tejlmand reviewed (changes requested):
lgtm.

One final comment I forgot last time.
Could you please remove this code, as this is now dead code when zephyr_module.py ensures there will be no duplicates anymore.
https://github.com/zephyrproject-rtos/zephyr/blob/24fd3d31f549c5a74d1cd0406e30f0a20ef0faf6/cmake/modules/zephyr_module.cmake#L149-L152

tejlmand reviewed (approved):
lgtm, thanks for this.

de-nordic reviewed (approved):
Looks OK.
